R/utils.R

```diff
@@ -71,4 +71,54 @@ defined_period_annual_growth <- function(data, begin, end){
   
   # Return new data frame
   period_growth
+}
+
+#=======================
+# Calculate tree density
+#=======================
+
+# This function calculates tree density as both the number of trees and the
+# summed area at breast height of trees within a user-defined distance of a
+# provided set of coordinates.
+
+nbhd_density <- function(mapping_data, stand, x, y, nbhd_radius){
+  
+  # Subset to stand of interest
+  focal_stand <- mapping_data %>%
+    filter(StandID == stand)
+  
+  # Create output table
+  x_coord <- NA
+  y_coord <- NA
+  num_trees <- NA
+  total_abh <- NA
+  output <- data.frame(x_coord, y_coord, num_trees, total_abh)
+  
+  # Begin looping through input coordinates
+  for(coord_num in 1:length(x)){
+    
+    focal_stand_summary <- focal_stand %>%
+      
+      # Calculate distance of each tree from input coordinates
+      mutate(dist = sqrt((abs(Xcoord - x[coord_num]) ^ 2) +
+                           (abs(Ycoord - y[coord_num]) ^ 2))) %>%
+      
+      # Subset to trees within input radius of input coordinates
+      filter(dist <= nbhd_radius) %>%
+      
+      # Calculate cross-sectional area at breast height (abh) for each tree
+      mutate(abh = pi * ((Dbh_cm / 2) ^ 2)) %>%
+      
+      # Extract number of trees and total abh within neighborhood
+      summarize(x_coord = x[coord_num],
+                y_coord = y[coord_num],
+                num_trees = n(),
+                total_abh = sum(abh))
```

Comment:
To get density, we'll need to divide by the area of the circle we are considering (i.e. divide by pi*100 for the default 10 meters)

Reply:
Function now divides by the neighborhood area